tiledb/cloud/taskgraphs/depgraph.py

```diff
@@ -0,0 +1,111 @@
+        parent_set = set(parents)
+        if child in parent_set:
+            raise CyclicGraphError(f"{child!r} can't be its own parent.")
+        missing_keys = parent_set - self._parent_to_children.keys()
+        if missing_keys:
+            raise KeyError(f"Entries {missing_keys!r} are not in the graph.")
+        # This isn't a no-op -- it ensures that the child node exists in both
+        # our mappings.
+        self._parent_to_children[child] = set()
+        self._child_to_parents[child] = set()
+        # When adding a completely new node to a DAG, you can't get a cycle.
+        for parent in parents:
+            self._add_edge_unsafe(child=child, parent=parent)
+        self._topo_sorted.append(child)
+
+    def add_edge(self, *, child: _T, parent: _T) -> None:
+        """Adds a new edge between two existing nodes."""
+        if child not in self._parent_to_children:
+            raise KeyError(f"{child!r} is not part of the graph")
+        if parent not in self._parent_to_children:
+            raise KeyError(f"{parent!r} is not part of the graph")
+        self._add_edge_unsafe(child=child, parent=parent)
+        try:
+            self._topo_sorted = self._topo_sort()
+        except CyclicGraphError:
+            self._remove_edge(child=child, parent=parent)
+            raise
+
+    def remove(self, node: _T) -> None:
+        """Removes a node, and all its connections, from the network."""
+        if node not in self._parent_to_children:
+            raise KeyError(f"{node!r} is not part of the graph")
+        for child in self._parent_to_children[node]:
+            self._child_to_parents[child].remove(node)
+        del self._parent_to_children[node]
+        for parent in self._child_to_parents[node]:
+            self._parent_to_children[parent].remove(node)
+        del self._child_to_parents[node]
+        self._topo_sorted.remove(node)
+
+    def parents_of(self, node: _T) -> FrozenSet[_T]:
+        """Returns the immediate parents of the given node."""
+        return frozenset(self._child_to_parents[node])
+
+    def children_of(self, node: _T) -> FrozenSet[_T]:
+        """Returns the immediate children of the given node."""
+        return frozenset(self._parent_to_children[node])
+
+    def _topo_sort(self) -> List[_T]:
+        in_degrees = {
+            node: len(parents) for node, parents in self._child_to_parents.items()
+        }
+        output: List[_T] = []
+        # All the nodes that have no parents are our root nodes.
+        q: Deque[_T] = collections.deque(
+            node for (node, deg) in in_degrees.items() if deg == 0
+        )
+        while q:
+            parent = q.popleft()
+            del in_degrees[parent]
+            for child in self._parent_to_children[parent]:
+                in_degrees[child] -= 1
+                if in_degrees[child] == 0:
+                    q.append(child)
+            output.append(parent)
+        if in_degrees:
```

Comment:
in general i don't like implicit truthiness. it's surely 'correct' or 'idiomatic' in some python-fluent perspective. but most of us are polyglots these days. it's important that we code in ways that are clearly correct to the average reader (even though you're a far-above-average writer when it comes to python).

personally i'd prefer `if len(in_degrees) == 0:` but i know there are always concerns about computing sizes of things when is-empty is all that's wanted. i'd be ok with a more explicit `if bool(in_degrees):` here.

p.s. did you know `if (0)` is true in ruby? my favorite example of why implicit truthiness is a Bad Thing. (and don't even get me started on what javascript lets people get away with ...)

Reply:
I'd prefer to keep using implicit `__nonzero__` here—it’s not only widely used in Python development but is regarded as a [good practice in the Google style guide](https://google.github.io/styleguide/pyguide.html#214-truefalse-evaluations) (if not elsewhere as well).